couchdb/prod.couchdb-cluster.yml

```diff
@@ -0,0 +1,68 @@
+version: "3.3"
+
+networks:
+  cht-net:
+    driver: bridge
+
+services:
+  couchdb.1:
+    image: medicmobile/cht-couchdb:${CHT_CORE_RELEASE:-clustered-042022-test31}
+    volumes:
+    - ${DB1_DATA:-./srv1}:/opt/couchdb/data
+    environment:
+      - "COUCHDB_USER=${COUCHDB_USER:-admin}"
+      - "COUCHDB_PASSWORD=${COUCHDB_PASSWORD:-password}"
+      - "COUCHDB_SECRET=${COUCHDB_SECRET:-6c1953b6-e64d-4b0c-9268-2528396f2f58}"
+      - "COUCHDB_UUID=${COUCHDB_UUID:-5c265815-b9e3-47f1-ba8d-c1d50495eeb2}"
+      - "SVC_NAME=${SVC1_NAME:-couchdb.1}"
+      - "CLUSTER_PEER_IPS=couchdb.2,couchdb.3"
+      - "COUCHDB_LOG_LEVEL=${COUCHDB_LOG_LEVEL:-error}"
+    restart: always
+    networks:
+      cht-net:
+
+  couchdb.2:
+    image: medicmobile/cht-couchdb:${CHT_CORE_RELEASE:-clustered-042022-test31}
+    volumes:
+    - ${DB2_DATA:-./srv2}:/opt/couchdb/data
+    environment:
+      - "COUCHDB_USER=${COUCHDB_USER:-admin}"
+      - "COUCHDB_PASSWORD=${COUCHDB_PASSWORD:-password}"
+      - "COUCHDB_SECRET=${COUCHDB_SECRET:-6c1953b6-e64d-4b0c-9268-2528396f2f58}"
+      - "COUCHDB_UUID=${COUCHDB_UUID:-5c265815-b9e3-47f1-ba8d-c1d50495eeb2}"
+      - "SVC_NAME=${SVC2_NAME:-couchdb.2}"
+      - "COUCHDB_LOG_LEVEL=${COUCHDB_LOG_LEVEL:-error}"
+    restart: always
+    networks:
+      cht-net:
+
+  couchdb.3:
+    image: medicmobile/cht-couchdb:${CHT_CORE_RELEASE:-clustered-042022-test31}
+    volumes:
+    - ${DB3_DATA:-./srv3}:/opt/couchdb/data
+    environment:
+      - "COUCHDB_USER=${COUCHDB_USER:-admin}"
+      - "COUCHDB_PASSWORD=${COUCHDB_PASSWORD:-password}"
+      - "COUCHDB_SECRET=${COUCHDB_SECRET:-6c1953b6-e64d-4b0c-9268-2528396f2f58}"
+      - "COUCHDB_UUID=${COUCHDB_UUID:-5c265815-b9e3-47f1-ba8d-c1d50495eeb2}"
+      - "SVC_NAME=${SVC3_NAME:-couchdb.3}"
+      - "COUCHDB_LOG_LEVEL=${COUCHDB_LOG_LEVEL:-error}"
+    restart: always
+    networks:
+      cht-net:
+
+  haproxy:
+    image: medicmobile/cht-haproxy:${CHT_CORE_RELEASE:-archv3-test6}
+    hostname: haproxy
+    environment:
+      - "HAPROXY_IP=${HAPROXY_IP:-haproxy}"
+      - "COUCHDB_USER=${COUCHDB_USER:-admin}"
+      - "COUCHDB_PASSWORD=${COUCHDB_PASSWORD:-password}"
+      - "COUCHDB1_SERVER=${COUCHDB1_SERVER:-couchdb.1}"
+      - "COUCHDB2_SERVER=${COUCHDB2_SERVER:-couchdb.2}"
+      - "COUCHDB3_SERVER=${COUCHDB3_SERVER:-couchdb.3}"
+      - "HAPROXY_PORT=${HAPROXY_PORT:-5984}"
+    ports:
+      - "${HAPROXY_PORT:-5984}:5984"
```

Comment:
I'm trying to run CHT-Core on top of this and I can't seem to log in (in the app). It's almost as if the session cookie from one node is not valid on another node?

Here is a screenshot of the network requests of me trying to log in: 

![image](https://user-images.githubusercontent.com/35681649/164488070-d36d9789-5216-46ea-99f9-4b4e2923ebfc.png)

You can see the login POST be successful (302) and then a request to `/medic` immediately responding with a 401.

Reply:
Ahh! 

Each node is generating a random salt when hashing the admin password. That salt is included in the session cookie, so the other nodes are denying you. **Nice catch!!** 

This is why our [test here](https://github.com/medic/cht-core/blob/couchdb-cluster-archv3-again/couchdb/tests/tests.bats#L52) passes with data being retrievable from any node, but logins across nodes with the same session is a different thing. I think we will need to refactor some of our automated cluster set-up, I'm thinking of some ways and will bring this up in our meeting today.

The other part of the session cookie is COUCHDB_SECRET, which we set the same as an env variable in the compose template.

Reply:
I've fixed this in the latest commit that exists inside this image: `medicmobile/cht-couchdb:clustered-042622-test37`. It's working across restarts and recreations. 

Question: Will we have to add the same logic for the service users to sync the hashed password across nodes?

Reply:
There are no service users. We abandoned them because creation required all nodes to be up.